src/js/load-package.ts

```diff
@@ -84,335 +90,437 @@ async function initializePackageIndex(lockFilePromise: Promise<Lockfile>) {
+        await nodeFsPromisesMod.mkdir(installBaseUrl, {
+          recursive: true,
+        });
+      }
+    } else {
+      installBaseUrl = API.config.indexURL;
     }
 
-    const { name: pkgname, version, fileName } = parsedPackageData;
-    const channel = name;
+    let fileName, uri, fileSubResourceHash;
+    if (pkg.channel === DEFAULT_CHANNEL) {
+      if (!(pkg.normalizedName in API.lockfile_packages)) {
+        throw new Error(`Internal error: no entry for package named ${name}`);
+      }
+      const lockfilePackage = API.lockfile_packages[pkg.normalizedName];
+      fileName = lockfilePackage.file_name;
 
-    if (toLoad.has(pkgname) && toLoad.get(pkgname)!.channel !== channel) {
-      errorCallback(
-        `Loading same package ${pkgname} from ${channel} and ${
-          toLoad.get(pkgname)!.channel
-        }`,
-      );
-      continue;
+      uri = resolvePath(fileName, installBaseUrl);
+      fileSubResourceHash = "sha256-" + base16ToBase64(lockfilePackage.sha256);
+    } else {
+      uri = pkg.channel;
+      fileSubResourceHash = undefined;
     }
-    toLoad.set(pkgname, {
-      name: pkgname,
-      normalizedName: pkgname,
-      channel: channel, // name is url in this case
-      depends: [],
-      installPromise: undefined,
-      done: createResolvable(),
-      packageData: {
-        name: pkgname,
-        version: version,
-        file_name: fileName,
+
+    if (!checkIntegrity) {
+      fileSubResourceHash = undefined;
+    }
+    try {
+      return await loadBinaryFile(uri, fileSubResourceHash);
+    } catch (e) {
+      if (!IN_NODE || pkg.channel !== DEFAULT_CHANNEL) {
+        throw e;
+      }
+    }
+    console.log(
+      `Didn't find package ${fileName} locally, attempting to load from ${this.cdnURL}`,
+    );
+    // If we are IN_NODE, download the package from the cdn, then stash it into
+    // the node_modules directory for future use.
+    let binary = await loadBinaryFile(this.cdnURL + fileName);
+    console.log(
+      `Package ${fileName} loaded from ${this.cdnURL}, caching the wheel in node_modules for future use.`,
+    );
+    await nodeFsPromisesMod.writeFile(uri, binary);
+    return binary;
+  }
+
+  /**
+   * Install the package into the file system.
+   * @param normalizedName The normalized name of the package
+   * @param buffer The binary data returned by downloadPackage
+   * @private
+   */
+  private async installPackage(
+    normalizedName: string,
+    buffer: Uint8Array,
+    channel: string,
+  ) {
+    let pkg = API.lockfile_packages[normalizedName];
+    if (!pkg) {
+      pkg = {
+        name: "",
+        version: "",
+        file_name: ".whl",
         install_dir: "site",
         sha256: "",
         package_type: "package",
-        imports: [],
+        imports: [] as string[],
```

Comment:
Is this needed? It knows the type of pkg...

Reply:
Right, typescript complained during the refactorings but it seems it is not needed anymore.